File: java/src/test/java/com/example/routeauthorizationtesting/RouteAuthSpec.java
package com.example.routeauthorizationtesting;

import org.springframework.http.HttpMethod;
import org.springframework.test.web.servlet.request.MockHttpServletRequestBuilder;

import java.util.Arrays;
import java.util.List;

import static java.util.stream.Collectors.toList;
import static org.springframework.security.test.web.servlet.request.SecurityMockMvcRequestPostProcessors.csrf;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;

public class RouteAuthSpec {
    String route;
    HttpMethod verb;
    Access access;

    RouteAuthSpec(String route, HttpMethod verb, Access access) {
        this.route = route;
        this.verb = verb;
        this.access = access;
    }

    MockHttpServletRequestBuilder getRequest() {
        String sanitizedRoute = route.replaceAll("\\{\\w+}", "0");
        switch (this.verb) {
            case GET:
                return get(sanitizedRoute);
            case POST:
                return post(sanitizedRoute).with(csrf());
            case PUT:
                return put(sanitizedRoute).with(csrf());
            case PATCH:
                return patch(sanitizedRoute).with(csrf());
            case DELETE:
                return delete(sanitizedRoute).with(csrf());
            default:
                throw new Error(String.format("Method %s not implemented", this.verb.toString()));
        }
    }
}

File: kotlin/src/test/kotlin/com/example/routeauthorizationtesting/RouteAuthSpec.kt
package com.example.routeauthorizationtesting

import org.springframework.http.HttpMethod
import org.springframework.http.HttpMethod.*
import org.springframework.security.test.web.servlet.request.SecurityMockMvcRequestPostProcessors.csrf
import org.springframework.test.web.servlet.request.MockHttpServletRequestBuilder
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*


data class RouteAuthSpec(
    val route: String,
    val verb: HttpMethod,
    val access: Access
) {
    val request: MockHttpServletRequestBuilder
        get() {
            val sanitizedRoute = route.replace(Regex("\\{\\w+\\}"), "0")
            return when (verb) {
                GET -> get(sanitizedRoute)
                POST -> post(sanitizedRoute).with(csrf())
                PUT -> put(sanitizedRoute).with(csrf())
                DELETE -> delete(sanitizedRoute).with(csrf())
                PATCH -> patch(sanitizedRoute).with(csrf())
                else -> TODO("Not Implemented for method $verb")
            }
        }

    override fun toString(): String {
        return "submits a $verb to $route"
    }
}

sealed class Access {
    object Unauthenticated : Access()
    class AnyRole(private vararg val roles: Role) : Access() {
        fun allowedForRole(role: Role): Boolean {
            return roles.contains(role)
        }
    }
}
